### PR TITLE
GEN-961: EIP 4626

### PR DESCRIPTION
**Pull request by fedealconada**
## Motivation

Sense wants to support [EIP-4626](https://eips.ethereum.org/EIPS/eip-4626) to benefit from the advantages of its standardisation. This way, integrations will be faster and safer since there won't be a need of implementing customised adapters (unless its preferred).  Later on, we could have a 4626 factory which allows anyone to deploy adapters whose target is an ERC4626.

## Solution

We've created the adapter `ERC4626Adapter.sol` whose target is an ERC4626. Moreover, we've adjusted our tests to also run over these kind of adapters.

## Implementer Checklist

- [ ]  [FIRST TIME ONLY] Review the [Solcurity](https://github.com/Rari-Capital/solcurity) standard
- [ ]  [OPTIONAL] Create a reference implementation (in python or JS)
- [x]  Create a short "Motivation" section for the PR
- [x]  Write a spec for the feature and put it in the PR description – basic function names and expected state transitions is OK
- [x]  Check all of the new revert paths with concrete tests
- [ ]  Check all of the new storage slot writes with concrete tests – do not make the passing test case the trivial case!
- [x]  Go line-by-line through the new code and ensure it has all been covered by concrete tests – since we don’t have a coverage engine for foundry yet, you can imagine how one might check that each different code path is covered
- [x]  Simplify the implementation and spend some time trying to minimize gas costs
- [x]  Write fuzz tests for the feature – try everything to break the implementation (is this function monotonically in/decreasing, should it always be less than something else, etc)
- [ ]  Capture bugs discovered during the above steps in concrete tests (regression tests)
- [x]  Add integration tests – fuzz or concrete tests that test how the new code affects the entire system, either locally or with a mainnet fork
- [ ]  [OPTIONAL] Integrate the feature into the deployment scripts
- [ ]  [OPTIONAL] Add new sanity checks to the deployment scripts
- [ ]  Get the PR reviewed by at least two people

### CODE REVIEW

**Comment by fedealconada** [comment]
need to fix some tests but it's ready for review @abylaw @jparklev

File: pkg/core/src/tests/adapters/CropAdapter.t.sol [hunk 1]
```diff
@@ -12,14 +12,23 @@ import { Divider } from "../../Divider.sol";
...
-import { TestHelper } from "../test-helpers/TestHelper.sol";
+import { TestHelper, MockTargetLike } from "../test-helpers/TestHelper.sol";
 
 contract CropAdapters is TestHelper {
     using FixedMath for uint256;
 
+    function setUp() public virtual override {
+        super.setUp();
+
+        // freeze scale to 1e18 (only for no 4626 targets)
```
Comment (jparklev) [inline pkg/core/src/tests/adapters/CropAdapter.t.sol]: So I can better understand, could you expand on the reasoning for this?

File: pkg/core/src/adapters/ERC4626CropsAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        uint128 _ifee,
+        AdapterParams memory _adapterParams,
+        address[] memory _rewardTokens
+    ) CropsAdapter(_divider, _target, address(ERC4626(_target).asset()), _ifee, _adapterParams, _rewardTokens) {
+        BASE_UINT = 10**ERC4626(target).decimals();
+        ERC20(underlying).approve(target, type(uint256).max);
+    }
+
+    function scale() external override returns (uint256) {
+        return ERC4626(target).convertToAssets(BASE_UINT);
```
Comment (jparklev) [inline pkg/core/src/adapters/ERC4626CropsAdapter.sol]: I've learned something about the expected usage of these functions since working on this last! Namely that some Targets are expected to have variable exchange rates depending on the size of the redemption. ie, the more you redeem, the worse price you get. And the `convertTo` functions are supposed to take that into account and return a conversion from one asset to another for the  “average user”

In the case of a variable ex rate, a single scale value doesn't work perfectly, but I think we can still get away with our pattern as long as we err on the conservative side. I'm thinking we do `convertToAssets` with the entire Target balance held in the adapter, then calculate the ex rate for one Target from there. Then, it's kind of a worst case exchange rate if we needed to unwrap all of the Target at once. Does this seem like the right idea to you?

Then also, `convertTo` converts to an asset balance, which will be in the decimals of the underlying in this case, and that could be something other than `18`, right?

File: pkg/core/src/tests/GYTManager.t.sol [hunk 3]
```diff
@@ -299,4 +301,23 @@ contract GYTsManager is TestHelper {
     //     assertEq(yieldBalanceAfter, aliceYieldBalance);
     //     assertTrue(tBalanceAfter > tBalanceBefore);
     // }
+
+    function calculateExcess(
```
Comment (jparklev) [inline pkg/core/src/tests/GYTManager.t.sol]: tbh we can probably just remove any files related to the gyt contract at this point. wdyt?

**Review by jparklev — COMMENTED** [review]
Haven't fully reviewed yet, just wanted to drop in a question and share something that i'd learned.

File: pkg/core/src/tests/adapters/CropAdapter.t.sol (hunk 1 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/CropAdapter.t.sol]: This was basically just a refactoring. All tests were having a `setScale(1e18)` so instead f having it on each test I moved it to the `setUp()` call.

File: pkg/core/src/adapters/ERC4626CropsAdapter.sol (hunk 2 above)
Comment (fedealconada) [reply pkg/core/src/adapters/ERC4626CropsAdapter.sol]: Great point, didn't think about this. So:
> Does this seem like the right idea to you?
yes, sounds good, so we should be doing something like:
```
uint256 assets = t.convertToAssets(t.totalSupply());
return assets.fdiv(t.totalSupply(), BASE_UINT)
```

> Then also, `convertTo` converts to an asset balance, which will be in the decimals of the underlying in this case, and that could be something other than 18, right?

Yes, it can. But this is already handled by the `ERC4626.sol` contract, right? Not sure if you are asking to double check or because you had a concern. The `ERC4626` implementation we're using (Rari's one) uses the underlying's decimals to set the target's decimals.

Comment (jparklev) [reply pkg/core/src/adapters/ERC4626CropsAdapter.sol]: > yes, sounds good, so we should be doing something like:

So I was thinking
```
uint256 targetBal = t.balanceOf(address(this);
uint256 assets = t.convertToAssets(targetBal));
return assets.fdiv(targetBal, BASE_UINT)
```

But I just realized there's an issue where a flashloan can change the scale value during it's execution 🙃 (since balanceOf would decrease during the loan). I think what you have makes sense, though I haven't thought about it too deeply yet and we might need circle back

> The ERC4626 implementation we're using (Rari's one) uses the underlying's decimals to set the target's decimals.

Yep, but we expect scale to always be in 18 decimals!

File: pkg/core/src/tests/adapters/CropAdapter.t.sol (hunk 1 above)
Comment (jparklev) [reply pkg/core/src/tests/adapters/CropAdapter.t.sol]: Right, I wonder if we should just always use the mock 4626 adapter by default to test other contracts (divider, periphery, etc), then we can test different ones for adapter-specific tests?

Comment (fedealconada) [reply pkg/core/src/tests/adapters/CropAdapter.t.sol]: Yeah, I thought about that but then just decided to have them both (ERC4626 and non-4626) running the whole test suite. My thought was more in the line of, the more tests we have the better. So we run all the test suite with one and with the other. Maybe, a little disadvantage of this approach is that I had to add some conditionals in several places to behave accordingly (you will in some places `if (is4626) ...`).

File: pkg/core/src/adapters/ERC4626CropAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    constructor(
+        address _divider,
+        address _target,
+        uint128 _ifee,
+        AdapterParams memory _adapterParams,
+        address _reward
+    ) CropAdapter(_divider, _target, address(ERC4626(_target).asset()), _ifee, _adapterParams, _reward) {
+        uint256 tDecimals = ERC4626(target).decimals();
+        BASE_UINT = 10**tDecimals;
+        SCALE_FACTOR = 10**(18 - tDecimals); // we assume targets decimals <= 18
```
Comment (abylaw) [inline pkg/core/src/adapters/ERC4626CropAdapter.sol]: should we add a explicit check for `tDecimals > 18` w/ a nice error message?

File: pkg/core/src/adapters/ERC4626Adapter.sol [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    constructor(
+        address _divider,
+        address _target,
+        uint128 _ifee,
+        AdapterParams memory _adapterParams
+    ) BaseAdapter(_divider, _target, address(ERC4626(_target).asset()), _ifee, _adapterParams) {
+        uint256 tDecimals = ERC4626(target).decimals();
+        BASE_UINT = 10**tDecimals;
+        SCALE_FACTOR = 10**(18 - tDecimals); // we assume targets decimals <= 18
```
Comment (abylaw) [inline pkg/core/src/adapters/ERC4626Adapter.sol]: same comment about explicitly checking that `tDecimals <= 18`

File: pkg/deployments/deploy/simulated/04-adapters.js [hunk 6]
```diff
@@ -209,12 +230,12 @@ module.exports = async function () {
...
   }
 
-  async function setScale(adapterAddress) {
+  async function setScale(adapterAddress, is4626) {
     const { abi: adapterAbi } = await deployments.getArtifact("MockAdapter");
     const adapter = new ethers.Contract(adapterAddress, adapterAbi, signer);
     const scale = await adapter.callStatic.scale();
     log(`-> scale: ${scale.toString()}`);
-    if (!scale.eq(ethers.utils.parseEther("1.1"))) {
+    if (!is4626 && !scale.eq(ethers.utils.parseEther("1.1"))) {
```
Comment (abylaw) [inline pkg/deployments/deploy/simulated/04-adapters.js]: nit: since this function is a noop when using a 4626 adapter, perhaps we want to rename it to `trySetScale`  instead? we could optionally log as well.

also, instead of passing in `is4626`, maybe we may want to check for `"setScale" in adapter` instead, which might be a little less error prone & is a bit simpler.

File: pkg/deployments/deploy/simulated/04-adapters.js [hunk 7]
```diff
@@ -151,17 +170,19 @@ module.exports = async function () {
     return await ethers.getContract(underlyingName, signer);
   }
 
-  async function deployTarget(targetName, tDecimals, underlyingAddress) {
+  async function deployTarget(targetName, tDecimals, underlyingAddress, is4626) {
```
Comment (abylaw) [inline pkg/deployments/deploy/simulated/04-adapters.js]: nit: I slightly prefer having two functions here:
`deployTarget(targetName, tDecimals, underlyingAddress)` & `deploy4626Target(targetName, underlyingAddress)` since they don't share that much code.

File: pkg/deployments/deploy/simulated/04-adapters.js [hunk 8]
```diff
@@ -76,24 +76,41 @@ module.exports = async function () {
   }
 
   // Helpers
-  async function deployAdapter(t, factory) {  
+  async function deployAdapter(t, factory, is4626) {
```
Comment (abylaw) [inline pkg/deployments/deploy/simulated/04-adapters.js]: nit: same comment as below, it may be easier to read to split into two functions, `deployAdapter` & `deploy4626Adapter`.

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 9]
```diff
@@ -5,18 +5,21 @@ import { ERC20 } from "@rari-capital/solmate/src/tokens/ERC20.sol";
...
+import { ERC4626CropAdapter } from "../../../adapters/ERC4626CropAdapter.sol";
+import { ERC4626CropsAdapter } from "../../../adapters/ERC4626CropsAdapter.sol";
 import { FixedMath } from "../../../external/FixedMath.sol";
 import { Divider } from "../../../Divider.sol";
 import { YT } from "../../../tokens/YT.sol";
 import { MockTarget } from "./MockTarget.sol";
 import { MockToken } from "./MockToken.sol";
+import { ERC4626 } from "@rari-capital/solmate/src/mixins/ERC4626.sol";
 
+// Mock crop adapter
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol:16]: nit: delete comment

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 10]
```diff
@@ -96,26 +128,33 @@ contract TestHelper is DSTest {
         inputs[1] = "_forge_mock_target_decimals";
         mockTargetDecimals = uint8(abi.decode(hevm.ffi(inputs), (uint256)));
 
+        inputs[1] = "_forge_mock_4626_target";
+        is4626 = uint8(abi.decode(hevm.ffi(inputs), (uint256))) == 0 ? false : true;
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol:132]:
```suggestion
        is4626 = uint8(abi.decode(hevm.ffi(inputs), (uint256))) != 0;
```

Comment (abylaw) [reply pkg/core/src/tests/test-helpers/TestHelper.sol:132]: also for my own education: what in the evm is happening here? 🤪

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 11]
```diff
@@ -96,26 +128,33 @@ contract TestHelper is DSTest {
...
-        }
-        if (mockTargetDecimals != 18) {
-            emit log_named_uint(
-                "Running tests with the mock Target token configured with the following number of decimals",
-                uint256(mockTargetDecimals)
-            );
+        // Log target setup
+        if (is4626) {
+            target = MockTargetLike(deployMockTarget(address(underlying), "Compound Dai", "cDAI", mockTargetDecimals));
+            emit log("Running tests with a 4626 mock target");
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol:142]:
```suggestion
            emit log(`Running tests with a 4626 mock target with ${mockTargetDecimals} decimals and mock underlying with ${mockUnderlyingDecimals} decimals`);
```
?

then we can delete the `emit`s after :)

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 12]
```diff
@@ -231,29 +255,55 @@ contract TestHelper is DSTest {
...
+        return _addLiquidityToBalancerVault(address(adapter), maturity, tBal);
+    }
+
+    function _addLiquidityToBalancerVault(
+        address adapter,
+        uint256 maturity,
+        uint256 tBal
+    ) internal {
+        MockTargetLike target = MockTargetLike(MockAdapter(adapter).target());
+        MockToken underlying = MockToken(!is4626 ? target.underlying() : target.asset());
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol:288]: nit: move this into if statement since it is only used within the `else`?

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 13]
```diff
@@ -264,16 +314,17 @@ contract TestHelper is DSTest {
             mode: MODE,
             tilt: 0
         });
-        someFactory = new MockFactory(address(divider), factoryParams, _reward); // deploy adapter factory
-        someFactory.addTarget(_target, true);
-        divider.setIsTrusted(address(someFactory), true);
-        periphery.setFactory(address(someFactory), true);
+        if (is4626) {
+            someFactory = address(new Mock4626CropFactory(address(divider), factoryParams, _reward)); // deploy 4626 crop adapter factory
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol]: nit: remove comment

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 14]
```diff
@@ -264,16 +314,17 @@ contract TestHelper is DSTest {
...
+        } else {
+            someFactory = address(new MockFactory(address(divider), factoryParams, _reward)); // deploy crop adapter factory
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol]: nit: remove comment (i think code is self-explanatory)

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 15]
```diff
@@ -284,43 +335,68 @@ contract TestHelper is DSTest {
             mode: MODE,
             tilt: 0
         });
-        someFactory = new MockCropsFactory(address(divider), factoryParams, _rewardTokens); // deploy adapter factory
-        someFactory.addTarget(_target, true);
-        divider.setIsTrusted(address(someFactory), true);
-        periphery.setFactory(address(someFactory), true);
+        if (is4626) {
+            someFactory = address(new Mock4626CropsFactory(address(divider), factoryParams, _rewardTokens)); // deploy 4626 crops adapter factory
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol]: nit: remove comment

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 16]
```diff
@@ -284,43 +335,68 @@ contract TestHelper is DSTest {
...
+        } else {
+            someFactory = address(new MockCropsFactory(address(divider), factoryParams, _rewardTokens)); // deploy crops adapter factory
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/TestHelper.sol]: nit: remove comment

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 17]
```diff
@@ -255,6 +312,57 @@ contract MockCropsAdapter is CropsAdapter {
...
+    function doCombine(uint256 maturity, uint256 uBal) external returns (uint256 tBal) {
+        tBal = Divider(divider).combine(address(this), maturity, uBal);
+    }
+
+    function doRedeemPrincipal(uint256 maturity, uint256 uBal) external {
+        Divider(divider).redeem(address(this), maturity, uBal);
+    }
+}
+
+// Mock base adapter
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol:396]: nit: remove comment

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 18]
```diff
@@ -255,6 +312,57 @@ contract MockCropsAdapter is CropsAdapter {
     }
 }
 
+// Mock ERC4626 crops adapter
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol:338]: nit: remove comment

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 19]
```diff
@@ -132,12 +135,66 @@ contract MockAdapter is CropAdapter {
...
+    function doCombine(uint256 maturity, uint256 uBal) external returns (uint256 tBal) {
+        tBal = Divider(divider).combine(address(this), maturity, uBal);
+    }
+
+    function doRedeemPrincipal(uint256 maturity, uint256 uBal) external {
+        Divider(divider).redeem(address(this), maturity, uBal);
+    }
+}
+
+// Mock crops adapter
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol:207]: nit: remove comment

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 20]
```diff
@@ -132,12 +135,66 @@ contract MockAdapter is CropAdapter {
     }
 }
 
+// Mock ERC4626 crop adapter
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol:145]: nit: remove comment

File: pkg/core/src/tests/Periphery.t.sol [hunk 21]
```diff
@@ -141,17 +142,23 @@ contract PeripheryTest is TestHelper {
             mode: MODE,
             tilt: 0
         });
-        MockFactory cropFactory = new MockFactory(address(divider), factoryParams, address(reward)); // deploy adapter factory
-        divider.setIsTrusted(address(cropFactory), true);
-        periphery.setFactory(address(cropFactory), true);
+
+        address cropFactory;
+        if (is4626) {
+            cropFactory = address(new Mock4626CropFactory(address(divider), factoryParams, address(reward))); // deploy 4626 crop adapter factory
```
Comment (abylaw) [inline pkg/core/src/tests/Periphery.t.sol]: nit: comments

File: pkg/core/src/tests/Periphery.t.sol [hunk 22]
```diff
@@ -1137,12 +1079,14 @@ contract PeripheryTest is TestHelper {
         assertEq(ptBal, 1e18);
     }
 
+    event Haha(uint256);
```
Comment (abylaw) [inline pkg/core/src/tests/Periphery.t.sol]: nit: 🤣

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    function testCantWrapMoreThanBalance() public {
+        uint256 wrapAmt = INITIAL_BALANCE + 1;
+
+        // Approvals
+        target.approve(address(erc4626Adapter), type(uint256).max);
+        underlying.approve(address(erc4626Adapter), type(uint256).max);
+
+        hevm.expectRevert("TRANSFER_FROM_FAILED");
```
Comment (abylaw) [inline pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:127]: is it possible to expect a generic revert? since we may not care too much about the specific error that caused the wrap/unwrap to fail?

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+        wrapAmt = bound(wrapAmt, 1, INITIAL_BALANCE);
+
+        // Approvals
+        target.approve(address(erc4626Adapter), type(uint256).max);
+        underlying.approve(address(erc4626Adapter), type(uint256).max);
+
+        // 1. Run a full wrap -> unwrap cycle
+        uint256 prebal = underlying.balanceOf(address(this));
+        uint256 targetFromWrap = erc4626Adapter.wrapUnderlying(wrapAmt);
+        assertEq(targetFromWrap, target.balanceOf(address(this)));
```
Comment (abylaw) [inline pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:72]: we may want to add a few sanity checks here (`targetFromWrap` > 0), otherwise `wrapUnderlying` & `unwrapTarget` could be noops & still pass the `assertEq(prebal, postbal)` below :)

ditto for part 4 of this test

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+            minm: MIN_MATURITY,
+            maxm: MAX_MATURITY,
+            mode: MODE,
+            tilt: 0,
+            level: Constants.DEFAULT_LEVEL
+        });
+
+        erc4626Adapter = new ERC4626Adapter(address(divider), address(target), ISSUANCE_FEE, adapterParams);
+    }
+
```
Comment (abylaw) [inline pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:61]: do we want to add a test for edge-case of unwrapping/wrapping 0 target/underlying? what is the expected behaviour there?

File: pkg/core/src/adapters/ERC4626CropAdapter.sol (hunk 4 above)
Comment (fedealconada) [reply pkg/core/src/adapters/ERC4626CropAdapter.sol]: in case someone wants to create an adapter in which the target passed is > 18 decimals, this subtraction will revert

File: pkg/core/src/tests/adapters/CropAdapter.t.sol (hunk 1 above)
Comment (jparklev) [reply pkg/core/src/tests/adapters/CropAdapter.t.sol]: A question is: what are we trying to test? Are we running the tests with two adapter types to test the components themselves (divider, periphery, etc), or is it to test the adapters. For the latter, I think I prefer to test have comprehensive adapter tests that we feel cover our bases, for the former I guess i'd just want us to be intentional about what we're testing

File: pkg/core/src/tests/Periphery.t.sol (hunk 22 above)
Comment (fedealconada) [reply pkg/core/src/tests/Periphery.t.sol]: lol

File: pkg/core/src/tests/Periphery.t.sol (hunk 21 above)
Comment (fedealconada) [reply pkg/core/src/tests/Periphery.t.sol]: what do you mean? :)

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol (hunk 23 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:127]: i don;t think there's a generic one but we kind of care about the error message (at least on this test). I would expect there that it reverts because of that specific error and not some other different one :)

File: pkg/core/src/tests/Periphery.t.sol (hunk 21 above)
Comment (abylaw) [reply pkg/core/src/tests/Periphery.t.sol]: ah, i think the comments can be removed since the code is fairly self-explanatory :)

Comment (fedealconada) [reply pkg/core/src/tests/Periphery.t.sol]: gotcha!

File: pkg/deployments/deploy/simulated/04-adapters.js (hunk 6 above)
Comment (fedealconada) [reply pkg/deployments/deploy/simulated/04-adapters.js]: i've changed this :) @abylaw

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol (hunk 25 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:61]: added!

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol (hunk 24 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:72]: added!

File: pkg/core/src/tests/adapters/CropAdapter.t.sol (hunk 1 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/CropAdapter.t.sol]: we've decided we're gonna keep the `is4626` at least for now and the idea running all the test suite with both a customised (non-erc4626) adapter and an erc4626 adapter.

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol (hunk 23 above)
Comment (abylaw) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:127]: Ahh, ok :) Mind clarifying a bit for me? I think my confusion stems from the fact that it's not immediately obvious to me why trying to wrap more than balance would lead to a transfer failure.

**Review by abylaw — APPROVED** [review]
LGTM! Thanks for the changes!!

File: pkg/core/src/tests/adapters/ERC4626Adapter.t.sol (hunk 23 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:127]: >  why trying to wrap more than balance would lead to a transfer failure.

this is because the adapter is first trying to make a `transferFrom` of the user's underlying balance into the adapter ([here](https://github.com/sense-finance/sense-v1/pull/250/files/ecd0348d36737488e87263f41546c8fb1c62e275#diff-4dc0b2e7ea69b908bba676964700f1033425d4a69a28d6b4ec6104e8ebeb1947R54)). As the user does not have that balance that the adapter is trying to pull from his "account", it triggers a `TRANSFER_FROM_FAILED`

Comment (abylaw) [reply pkg/core/src/tests/adapters/ERC4626Adapter.t.sol:127]: Ahhhh, ok, that makes sense to me now :)

**Comment by fedealconada** [comment]
approved @jparklev ?

**Review by jparklev — APPROVED** [review]
approved
